src/black/strings.py

```diff
@@ -236,3 +238,57 @@ def normalize_string_quotes(s: str) -> str:
         return s  # Prefer double quotes
 
     return f"{prefix}{new_quote}{new_body}{new_quote}"
+
+
+@lru_cache(maxsize=4096)
+def char_width(char: str) -> int:
+    """Return the width of a single character as it would be displayed in a
+    terminal or editor (which respects Unicode East Asian Width).
+
+    Full width characters are counted as 2, while half width characters are
+    counted as 1.  Also control characters are counted as 0.
+    """
+    table = WIDTH_TABLE
+    codepoint = ord(char)
+    highest = len(table) - 1
+    lowest = 0
+    idx = highest // 2
+    while True:
+        start_codepoint, end_codepoint, width = table[idx]
+        if codepoint < start_codepoint:
+            highest = idx - 1
+        elif codepoint > end_codepoint:
+            lowest = idx + 1
+        else:
+            return 0 if width < 0 else width
+        if highest < lowest:
+            break
+        idx = (highest + lowest) // 2
+    return 1
+
+
+def str_width(line_str: str) -> int:
+    """Return the width of `line_str` as it would be displayed in a terminal
+    or editor (which respects Unicode East Asian Width).
+
+    You could utilize this function to determine, for example, if a string
+    is too wide to display in a terminal or editor.
+    """
+    if COMMON_SINGLE_WIDTH_CHARS.match(line_str):
```

Comment:
The reason I suggested `.isascii()` specifically is that thanks to a peculiarity of how strings are implemented, it can just look at an internal flag on the string object and so it takes O(1) time. Therefore, it's significantly faster than other similar-seeming methods like `isalpha()`, which in turn are quite a bit faster than running the regex engine.

To demonstrate:
```
In [3]: COMMON_SINGLE_WIDTH_CHARS: Final = re.compile(r"^[\x20-\x7e\xa0-\u02ff]*$")

In [4]: s = "a" * 40

In [5]: %timeit COMMON_SINGLE_WIDTH_CHARS.match(s)
229 ns ± 0.664 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [6]: %timeit s.isascii()
11.5 ns ± 0.0159 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [7]: %timeit s.isalpha()

79.2 ns ± 0.102 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

So `isascii` is about 20x faster than this regex.

If we use `isascii`, we'll no longer identify control chars as width 0. But I think that might be the right call anyway, because many editors will display control-chars (e.g. here is VSCode on a file with all the ASCII-range characters):

<img width="249" alt="Screen Shot 2023-01-20 at 4 27 06 AM" src="https://user-images.githubusercontent.com/906600/213694759-56f4fa44-378a-4dfb-846a-ccb689048fa0.png">

So I think we should stick with `isascii`.